app/rust/src/lib.rs

```diff
@@ -77,102 +87,71 @@ pub extern "C" fn scalar_multiplication(
-    start_index: &mut [u8; 11],
-    d_l: &mut [u8; 44],
-) -> ParserError {
-    let start = &mut *start_index;
-    let diversifier =  &mut *d_l;
-    get_diversifiers(dk,  start, diversifier);
-    ParserError::ParserOk
-}
-
-#[no_mangle]
-pub extern "C" fn is_valid_diversifier(
-    hash: &[u8; 32],
-) -> bool {
+pub extern "C" fn is_valid_diversifier(hash: &[u8; 32]) -> ParserError {
     let u = AffinePoint::from_bytes(*hash);
-    if u.is_some().unwrap_u8() == 1 {
-        let q = u.unwrap().mul_by_cofactor();
-        return q != ExtendedPoint::identity();
+
+    // Check if the CtOption is Some
+    if u.is_some().into() {
+        // Convert CtOption to bool
+        let point = u.unwrap(); // Safe to unwrap since we checked is_some
+        let q = point.mul_by_cofactor(); // Use the point directly
+        if q != ExtendedPoint::identity() {
+            return ParserError::ParserOk; // Valid diversifier
+        }
     }
 
-    false
+    ParserError::ParserUnexpectedError // Return error if the point is not valid
 }
 
 #[no_mangle]
-pub extern "C" fn get_pkd(
-    ivk_ptr: &[u8; 32],
-    h: &[u8; 32],
-    pk_d: &mut [u8; 32],
+pub extern "C" fn randomized_secret_from_seed(
+    ask: &[u8; 32],
+    alpha: &[u8; 32],
+    output: &mut [u8; 32],
 ) -> ParserError {
+    let skfr = Fr::from_bytes(ask);
+    if skfr.is_none().into() {
+        return ParserError::ParserUnexpectedError; // Handle error for ask
+    }
+    let skfr = skfr.unwrap(); // Safe to unwrap since we checked is_none
 
-    let affine = AffinePoint::from_bytes(*h).unwrap();
-    let extended = ExtendedPoint::from(affine);
-    let cofactor = extended.mul_by_cofactor();
-    let p = cofactor.to_niels().multiply_bits(ivk_ptr);
-    *pk_d = AffinePoint::from(p).to_bytes();
-
-    ParserError::ParserOk
-}
-
-#[no_mangle]
-pub extern "C" fn randomized_secret_from_seed(
-    ask:  &[u8; 32],
-    alpha:  &[u8; 32],
-    output:  &mut [u8; 32],
-) -> ParserError{
+    let alphafr = Fr::from_bytes(alpha);
+    if alphafr.is_none().into() {
+        return ParserError::ParserUnexpectedError; // Handle error for alpha
+    }
+    let alphafr = alphafr.unwrap(); // Safe to unwrap since we checked is_none
 
-    let mut skfr = Fr::from_bytes(ask).unwrap();
-    let alphafr = Fr::from_bytes(alpha).unwrap();
-    skfr += alphafr;
-    output.copy_from_slice(&skfr.to_bytes());
+    let skfr_result = skfr + alphafr;
+    output.copy_from_slice(&skfr_result.to_bytes());
 
     ParserError::ParserOk
 }
 
 #[no_mangle]
 pub extern "C" fn compute_sbar(
-    s:  &[u8; 32],
-    r:  &[u8; 32],
-    rsk:  &[u8; 32],
-    sbar:  &mut [u8; 32],
-) -> ParserError{
-    let s_point = Fr::from_bytes(s).unwrap();
-    let r_point = Fr::from_bytes(r).unwrap();
-    let rsk_point = Fr::from_bytes(rsk).unwrap();
+    s: &[u8; 32],
+    r: &[u8; 32],
+    rsk: &[u8; 32],
+    sbar: &mut [u8; 32],
+) -> ParserError {
+    let s_point = Fr::from_bytes(s);
+    if s_point.is_none().into() {
```

Comment:
same here, the is_none().into()
maybe i am missing something and is_none() here does not return a boolean, because it is not a core::Option type?